File: src/commands/categories.ts
import * as commander from 'commander';
import Table from 'cli-table';

import { actionRunner } from '../utils.js';
import { getAccessToken } from '../config.js';
import { getCategoriesForBenefitName } from '../forma.js';
import VERSION from '../version.js';

const command = new commander.Command();

interface Arguments {
  accessToken?: string;
  benefit: string;
}

command
  .name('categories')
  .version(VERSION)
  .description('List categories available for a Forma benefit')
  .requiredOption('--benefit <benefit>', 'The benefit to list categories for')
  .option('--access-token <access_token>', 'Access token used to authenticate with Forma')
  .action(
    actionRunner(async (opts: Arguments) => {
      const accessToken = opts.accessToken ?? getAccessToken();

      if (!accessToken) {
        throw new Error(
          "You aren't logged in to Forma. Please run `npx formanator login` first.",
        );
      }

      const categories = await getCategoriesForBenefitName(accessToken, opts.benefit);

      const table = new Table({
        head: ['Parent Category', 'Category'],
      });

      for (const category of categories) {
        table.push([
          category.category_name,
          category.subcategory_alias ?? category.subcategory_name,
        ]);
      }

      console.log(table.toString());
    }),
  );

export default command;

File: src/commands/categories.rs
use anyhow::Result;
use tabled::{Table, Tabled, settings::Style};

use crate::cli::CategoriesArgs;
use crate::config::resolve_access_token;
use crate::forma::get_categories_for_benefit_name;
use crate::verbose;

#[derive(Tabled)]
struct Row {
    #[tabled(rename = "Parent Category")]
    parent: String,
    #[tabled(rename = "Category")]
    category: String,
}

pub fn run(args: CategoriesArgs) -> Result<()> {
    verbose::set(args.verbose);
    let access_token = resolve_access_token(args.access_token.as_deref())?;
    let categories = get_categories_for_benefit_name(&access_token, &args.benefit)?;

    let rows: Vec<Row> = categories
        .into_iter()
        .map(|c| Row {
            parent: c.category_name,
            category: c.subcategory_alias.unwrap_or(c.subcategory_name),
        })
        .collect();

    let mut table = Table::new(rows);
    table.with(Style::sharp());
    println!("{table}");
    Ok(())
}
